website/src/hooks/tasks/useGenericTaskAPI.tsx

```diff
@@ -1,26 +1,43 @@
 import { useState } from "react";
 import { get, post } from "src/lib/api";
-import { BaseTask, TaskResponse, TaskType as TaskTypeEnum } from "src/types/Task";
+import { TaskApiHook } from "src/types/Hooks";
+import { BaseTask, TaskAvailableResponse, TaskResponse, TaskType as TaskTypeEnum } from "src/types/Task";
 import useSWRImmutable from "swr/immutable";
 import useSWRMutation from "swr/mutation";
 
-export const useGenericTaskAPI = <TaskType extends BaseTask>(taskType: TaskTypeEnum) => {
-  type ConcreteTaskResponse = TaskResponse<TaskType>;
+export const useGenericTaskAPI = <TaskType extends BaseTask>(taskType: TaskTypeEnum): TaskApiHook<TaskType> => {
+  const [response, setReponse] = useState<TaskResponse<TaskType>>({ status: "AWAITING_INITIAL" });
+  const [isLoading, setIsLoading] = useState(true);
 
-  const [tasks, setTasks] = useState<ConcreteTaskResponse[]>([]);
+  const { mutate: requestNewTask } = useSWRImmutable<TaskAvailableResponse<TaskType>>(
+    "/api/new_task/" + taskType,
+    get,
+    {
+      onSuccess: (response) => {
+        setIsLoading(false);
+        setReponse({ status: "AVAILABLE", ...response });
```

Comment:
could you move the status to its own variable? no need to pollute the actual response.

Reply:
I'm inclined to think that if we get a `200` the task should be assumed available, considering if it's not we get an error and a corresponding error code.

Reply:
The intent was to make it so that when you have a TaskAvailable you have a task, but if you have an AwaitingInitialTask or NoTaskAvailable you don't. If we separate the variables we are back in the situation where the variables only update one at a time and we would have to handle the cases that you have the "AVAILABLE" state but no task and the "NONE_AVAILABLE" state yet have a task. Originally the loading was also part of the tagged union `TaskResponse` but in testing I (thought I) discovered that we actually intend the isLoading to be possible regardless of the current state of having a task.

Reply:
just to make sure I understood: we use one variable instead of two because they are updated independently, and react might render in-between the updates, leading to an undefined behaviour?

Reply:
That is part of it, but the other advantage is it reflects reality more accurately, by definition if the task is available we have task data and when it's not we don't, reflecting that in the types just leads to code that is only as complex as reality.

Reply:
I see what you mean, you can set the `task` to `null` while loading, this will also reflect the reality as accurately as before.

The reason they should be separate is to separate the concerns, the status is only used in the `TaskPage`.

Regarding:
> reflecting that in the types just leads to code that is only as complex as reality.

if we want to have an interface for every possible state our system could be in, we will end up with exponentially many, which is unmaintainable in the long run.